src/function.ts

```diff
@@ -44,17 +57,154 @@ export type FunctionClosure<P, O> = (
   context: Context
 ) => Promise<O>;
 
-export interface IFunction<P, O> {
+/**
+ * Returns the payload type on the {@link IFunction}.
+ */
+export type FunctionPayload<F extends IFunction<any, any>> = [F] extends [IFunction<infer P, any>] ? P : never;
+/**
+ * Returns the output type on the {@link IFunction}.
+ */
+ export type FunctionOutput<F extends IFunction<any, any>> = [F] extends [IFunction<any, infer O>] ? O : never;
```

Comment:
nit: indentation

Reply:
why the `[F] extends `? What's the purpose of packing it a 1-tuple?

Reply:
`[F] extends []` changes the way that union types are interpreted.

https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types

With `[F]`

![image](https://user-images.githubusercontent.com/289213/172044197-5c567581-06ed-47b9-a030-4643409a967b.png)

Without

![image](https://user-images.githubusercontent.com/289213/172044243-98db4c3c-e616-4e64-a36a-1ac60535b6bb.png)

Which in the case its being used in... if the types are not blown out (`Function<A | B>`) then the `isFunction` does not narrow the types correctly. 

```ts
if(isFunction(destination)) {
    destination // is IFunction<AsyncResponseSuccess<P, O> | AsyncResponseFailure<P>, any>
} else {
     destination // is IDestination | IFunction<AsyncResponseSuccess<P, O> | AsyncResponseFailure<P>, any>
}
```

![image](https://user-images.githubusercontent.com/289213/172044280-d5638b53-48cc-40a4-af8e-1a0341166568.png)

But with... we get nice type narrowing.

```ts
if(isFunction(destination)) {
    destination // is IFunction<AsyncResponseSuccess<P, O>, any> | IFunction<AsyncResponseFailure<P>, any>
} else {
     destination // is IDestination
}
```

![image](https://user-images.githubusercontent.com/289213/172044366-e71f6af0-06a5-4d84-905c-8c256d71ac71.png)

Reply:
Awesome analysis. Thanks!